### PR TITLE
feat(tracer): Register variables in trace

### PR DESCRIPTION
**Pull request by smanilov**
First commit: the foundation is laid here to register variables during execution (debugging) of a Noir program.
Second commit: converts UnsignedInteger in addition to Field, and restores the tests deleted in the previous commit.

### CODE REVIEW

File: tooling/tracer/src/debugger_glue.rs [hunk 1]
```diff
@@ -42,8 +42,9 @@ fn convert_debugger_stack_frame(
     debugger_stack_frame: &noirc_artifacts::debug::StackFrame<FieldElement>,
 ) -> StackFrame {
     let function_name = String::from(debugger_stack_frame.function_name);
-    // TODO(stanm): add call params and values
-    StackFrame { function_name }
+    // TODO(stanm): add call params
+    let variables = debugger_stack_frame.variables.iter().map(Variable::from_tuple).collect();
+    StackFrame { function_name, variables }
```
Comment (spartakvarna) [inline tooling/tracer/src/debugger_glue.rs:47]: Does the function convert from StackFrame to StackFrame?

File: tooling/tracer/src/tracer_glue.rs [hunk 2]
```diff
@@ -11,6 +14,36 @@ pub(crate) fn register_step(tracer: &mut Tracer, location: &SourceLocation) {
...
+/// Registers a variable for the last registered step. Each time a new step is registered, all of
+/// its variables need to be registered too. If no variables are registered for a step, the
+/// frontend will not carry over the variables registered for the previous step.
+pub(crate) fn register_variable(tracer: &mut Tracer, variable: &Variable) {
+    let value_record = register_value(tracer, &variable.value, &variable.typ);
+    tracer.register_variable_with_full_value(&variable.name, value_record);
+}
+
+/// Registers a value of a given type. Registers the type, if it's the first time it occurs.
+fn register_value(
```
Comment (spartakvarna) [inline tooling/tracer/src/tracer_glue.rs:26]: lgtm

File: tooling/tracer/src/debugger_glue.rs (hunk 1 above)
Comment (smanilov) [reply tooling/tracer/src/debugger_glue.rs:47]: It does, but between different `StackFrame`s. From `noirc_artifacts::debug::StackFrame` to `crate::stack_frame::StackFrame`.

Any suggestions?

**Review by smanilov — COMMENTED** [review]
The test was actually not passing, because the `variables` array was not stable in the debugger.

So I made a tiny change to the first commit (added .sort() and the impl of the comparison traits in `stack_frame.rs`).

There is also now a new commit, which reintroduces the previously failing tests. They work again, because the code also handles `UnsignedInteger`, in addition to `Field`.

PTAL

File: tooling/tracer/src/tracer_glue.rs [hunk 3]
```diff
@@ -11,6 +15,68 @@ pub(crate) fn register_step(tracer: &mut Tracer, location: &SourceLocation) {
...
+        }
+        PrintableType::UnsignedInteger { width } => {
+            if let PrintableValue::Field(field_value) = value {
+                let mut noir_type_name = String::new();
+                if let Err(err) = write!(&mut noir_type_name, "u{width}") {
+                    panic!("failed to generate Noir type name: {err}");
+                }
+                let type_id =
+                    tracer.ensure_type_id(runtime_tracing::TypeKind::Int, &noir_type_name);
+                ValueRecord::Int { i: field_value.to_i128() as i64, type_id }
```
Comment (alehander92) [inline tooling/tracer/src/tracer_glue.rs:65]: can there be 128-bit ints in noir? should we support them?

Comment (smanilov) [reply tooling/tracer/src/tracer_glue.rs:65]: The default Field value is something like "254 bits", but not something we should worry about at the moment.

**Pull request merged**